### ISSUE TITLE
Browser "view source" exposes Backdrop and other version numbers when js aggregation is turned off

### ISSUE THREAD

**Issue by swampopus** [backdrop/backdrop-issues#7179]
This isn't exactly a bug or even a security issue, so I wasn't sure how to label it or what kind of template to select for this issue.

# The Problem #

If you have JS Aggregation turned off, and click to "view source" of any page, even as an anonymous user, you can clearly see the Backdrop core version number.

See example:

<img width="925" height="675" alt="Image" src="https://github.com/user-attachments/assets/b8ef5f66-d8f4-4086-86d8-b27e7be7aad0" />

This has some future security implications, as if there's ever an exploit discovered that targets, for example, anything before version 1.50.2, then it would be that much easier for a bad actor to find Backdrop sites that were exploitable.

# The Fix #

Instead of adding the version numbers as part of the URL, they should get the same cache-break string added as the contrib CSS and JS files.

Ex:   `<script src="https://example.com/modules/contrib/devel/devel_krumo.js?t3t45p"></script>`

The `t3t45p` is what I'm talking about at the end.  It changes every time the Backdrop cache is cleared, which is fine while you're actively developing and working on your site.  

If you have "aggregate JS / CSS" turned on, this occludes the version numbers.

# Alternatives #

As an alternative, on the Performance form, where you can disable JS and CSS aggregation, have a warning that if it is disabled, then there are some security implications.  But to me this seems like it would lead to more confusion.

**Comment by swampopus** [comment]
So the PR failed some tests specifically because they were looking for the version strings on the JS files.  And my thesis is that having the version strings there is a bad idea, and shouldn't be visible.

Not sure where to go from here!  Feedback is welcome.

EDIT:  I looked up some random Drupal 10 sites, and they're still showing the exact Drupal version in production:

<img width="371" height="121" alt="Image" src="https://github.com/user-attachments/assets/402af575-7e53-493b-b7e6-e250bc14b138" />

I feel like I shouldn't mention what site this is, but it was listed on a "top drupal 10 sites" list.  Maybe this isn't as big of a deal as I think it is?  Just seems like asking for trouble though.

EDIT 2: This D10 site is an ecommerce site and _allegedly_ makes $500M yearly revenue.  I feel like I should tell them that I can see their version number is 10.6.10?  There have been some security releases since then, including a "critical" one which apparently involved PHP Object injection vulnerability (SA-CORE-2026-005).  But this kind of underscores my reasoning for why we shouldn't be showing people our versions either.

**Comment by swampopus** [comment]
@backdrop/core-committers Hello core committers.  So this PR removes version numbers from the final html output on sites with "aggregate js" turned off.  It's for hardening security.  The trouble is that it fails 3 tests which are specifically looking for those version numbers.  I wasn't sure how to proceed at this point.

**Comment by olafgrabienski** [comment]
@swampopus – Thanks for the PR. I'm not able to do a code review, but I've applied the PR on a test site, and it works like a charm for me!

**Comment by argiepiano** [comment]
I'm doubtful that the PR is a good idea. The `v=...` was [added](https://www.drupal.org/project/drupal/issues/721400) there for a reason, and the fact that there are tests to test this functionality tells me that this is important. What happens when we update jquery to the next version? Or if we update any of the js libraries shipped with Backdrop? Browser caches will still use an old version of the libraries (no cache busting). Additionally the D7 issue seems to indicate that, before the `v=` query string, D7 was too aggressively rebuilding JS aggregation. The appended version was a good solution to do both things: cache busting when changing versions, and allowing aggregation without rebuilding everything on every cache flush.

Perhaps there is a way to replace the version number with something else, to finger-print the the files without revealing the version? 

EDIT: I realize now that you are keeping the default query string as a way to do cache busting - but if I understand the D7 issue correctly, this will change on every cache clear, which means doing a full-fledged aggregation every time. I think the goal of the D7 change was to avoid this?

**Comment by swampopus** [comment]
@argiepiano I'm willing to accept that the version numbers are there for some reason, I just can't find an explanation as to why.  And in fact, the Internet is filled with articles about why you should never reveal your version numbers.  ChatGPT inspected one of my sites and even told me it wasn't secure because it could see version numbers.  And in my example earlier in this post, I found a live D10 site with a critical bug I could attempt to exploit because I know what the version number is.

Correct me if I am wrong, but if you update jquery or any other js file that ships with Backdrop core, it means you are going to end up clearing the cache, right?  Due to the upgrade process itself?  As soon as you do, you get a new cache-bust string and everything is reloaded fresh again.  The browser will correctly download the new versions.

The Drupal issue you posted has 220+ comments :).  I'm too old to read through it all-- what exactly is in there explains why they needed to add version numbers?

Re: a "fingerprint" of the file replacing the version number-- that's very similar to what my PR does.  It just changes every time you clear cache.  And if you have JS Aggregation enabled, then Backdrop (and Drupal) do not append version numbers by default.

**Comment by swampopus** [comment]
Also-- I just did a quick test, and with Aggregation turned ON, it does NOT rebuild the aggregated js files with each cache flush.  So there's no issue there.

**Comment by quicksketch** [comment]
From @swampopus's original screenshot:

> Also, what version of jQuery is showing as well.

I'm not sure there's much we can do about library versions. Since we want to be using the official distributed versions, they're really easy to fingerprint just by looking at the loaded files. The version is right there at the top of the file. That's true for most 3rd party libraries.

From @argiepiano:

> I realize now that you are keeping the default query string as a way to do cache busting - but if I understand the D7 issue correctly, this will change on every cache clear, which means doing a full-fledged aggregation every time. 

The change @swampopus is requesting should only affect non-aggregated files.

---

On the whole, the big change here is that browser caches will get cleared for all non-aggregated core and contrib JS/CSS files if you use the Flush all Caches button or admin bar option. Previously it would only clear the browser cache for files that did not specify a version, which would typically only apply to custom code.

I could go either way on this. Right now a full cache clear doesn't actually clear the browser cache for non-aggregated core and contrib JS/CSS files; and maybe it should. Using the library version was a convenient way to only clear the browser cache when a change occurred within that file by assuming that the version will change when the file contents change. I've been tripped up by that a few times when doing development that the browser cache persists if I'm working on a dev version (i.e. `?v=1.34.x`) because the file contents might change while I'm working on it but cache clears do not update the query string.

Overall, we're talking about a small performance penalty (less browser caching for unaggregated files) for a small security hardening (slightly concealing the installed version numbers).

**Comment by avpaderno** [comment]
Not explicitly showing the version would be possible by using `hash('crc32', $version)` instead of `$version`. That would still make possible to understand which version is used for a library, or to notice when a different library version is used.  
Using `hash('murmur3f', $version, options: ['seed' => $seed])` would make possible to get a value that does not uniquely depend on `$version`, but that would require PHP 8.1.0, and more code to handle `$seed`, store it in some place, and decide when it should be changed.
The change proposed here is the minimal change to avoid showing library versions in the page source.

I would avoid showing the library versions in the page source for the same reason Backdrop core adds `<meta name="Generator" content="Backdrop CMS 1 (https://backdropcms.org)" />` instead of adding `<meta name="Generator" content="Backdrop CMS 1.33.1 (https://backdropcms.org)" />`. (For Backdrop libraries, the page source actually shows the exact Backdrop version used by a site, when CSS/JavaScript aggregation is not used.)

**Comment by quicksketch** [comment]
@avpaderno That is an interesting idea. I haven't heard of one-way hashes that also take a seed. We could potentially use the already existing `hash_salt` settings.php variable and only use `murmur3f` if on PHP 8.1 or higher, and use `crc32` otherwise. That would obscure the version minimally on older versions of PHP and "adequately" on newer versions; all while keeping the current behaviors.

**Comment by swampopus** [comment]
It's not a bad idea, and if that's what it takes to get consensus, I'd be happy to implement it.  But this wouldn't fix the issue you mentioned @quicksketch , with tripping you up in development of new library versions that still have the old version numbers set.  And you're right-- a tiny performance hit on cache clears, on dev sites, is offset by security hardening.  Also, adding hashes adds a bit more overhead and complexity to the code.  I know it's tiny, but since we already have the cache busting string anyway, I'd argue to just use that for readability and simplicity.

But my main goal is to get rid of the version nums.  If you guys want me to put in a hash of the version num instead, I'm happy to do it.  

Perhaps instead of worrying about a seed (and having to check version num), it instead uses the existing site token with the version number as the salt:

```php
$cache_buster = backdrop_get_token($version);
```
The site's token does not change per cache flush, and is still private enough for this application.

**Comment by avpaderno** [comment]
The value returned by [`backdrop_get_token()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_get_token/1) depends on the session ID; it is also a cryptographic hash, which means it is a too long value for this purpose.
Since it depends on the session ID, the browser cache would be bypassed at every request, for an anonymous user access.

**Comment by swampopus** [comment]
Ah, yeah that's true.  Well then a simple hash of version + salt would work:
```php
$cache_buster = md5($version . backdrop_get_hash_salt());
```
And that's if this really even necessary. I still don't think it is, but I'm happy to change my PR if this is what is wanted.

**Comment by avpaderno** [comment]
MD5 is a (broken) cryptographic hash. New code should either use a different cryptographic hash or use a non-cryptographic hash. #5785 has been open to replace any usage of MD5 in Backdrop core.

A non-cryptographic hash is necessary to hide the library version but still have a value that is created from the library version. In that way, the browser cache is only bypassed a new library version is used, but looking at the page source would not reveal the currently used version (or the exact Backdrop core version).

The change is minimal.

```php
case 'file':
  if (!empty($item['version'])) {
    $hash_algos = hash_algos();
    $hash_salt = backdrop_get_hash_salt();

    if (in_array('murmur3f', $hash_algos, TRUE)) {
      $seed = hexdec(substr($hash_salt, 0, 16));
      $hash = hash('murmur3f', $item['version'], FALSE, array('seed' => $seed));
    }
    else {
      $seed = substr($hash_salt, 0, 8);
      $hash = hash('crc32', $item['version'] . $seed);
    }

    $query_string = "v=$hash";
  }
  else {
    $query_string = $default_query_string;
  }

  $query_string_separator = (strpos($item['data'], '?') !== FALSE) ? '&' : '?';
  $element['#attributes']['src'] = file_create_url($item['data']) . $query_string_separator . ($item['cache'] ? $query_string : REQUEST_TIME);
  break;
```

Using xxHash, the code would be the following one.

```php
case 'file':
  if (!empty($item['version'])) {
    $hash_algos = hash_algos();
    $hash_salt = backdrop_get_hash_salt();

    if (in_array('xxh128', $hash_algos, TRUE)) {
      $segret = substr($hash_salt, 0, 136);
      $hash = hash('xxh128', $item['version'], FALSE, array('segret' => $segret));
    }
    else {
      $seed = substr($hash_salt, 0, 8);
      $hash = hash('crc32', $item['version'] . $seed);
    }

    $query_string = "v=$hash";
  }
  else {
    $query_string = $default_query_string;
  }

  $query_string_separator = (strpos($item['data'], '?') !== FALSE) ? '&' : '?';
  $element['#attributes']['src'] = file_create_url($item['data']) . $query_string_separator . ($item['cache'] ? $query_string : REQUEST_TIME);
  break;
```

(I corrected a typo in a PHP function name.)

_murmur3f_ and _xxh128_ are both 128-bit hashes.

**Comment by avpaderno** [comment]
As a side note, since MurmurHash and xxHash are not cryptographic hashes, they could return the same value for different inputs. Given the input value is a version number, and not a completely random string, that risk should be reduced.

**Comment by swampopus** [comment]
Heh, I know it's not cryptographically secure.  I'm just trying to get a unique string with as few steps (and characters) as possible.  I don't even think it's needed.  But for what it's worth, Backdrop would have to put out _literally billions_ of versions before there's a _potential_ collision of the salted, hashed version string with md5().

I would rather just not have a hashed version string at all.  What problem is it solving that the default query string doesn't?

**Comment by swampopus** [comment]
Tell ya what-- I'm just going to leave my PR where it is.  If anyone else would like to make a more complex hashing scheme to keep the version numbers obfuscated yet different from the default query string, please feel free.  I'll leave it to the community.

**Comment by avpaderno** [comment]
> What problem is it solving that the default query string doesn't?

The value of `$default_query_string` is changed every time the cache is cleared from Backdrop side; the values returned by `hash('crc32', $item['version'])` and `hash('murmur3f', $item['version'])` change only when `$item['version']` is changed.
I think that one of the objections given in a comment was that the browser cache is bypassed too often, when `$default_query_string` is always used.

**Comment by avpaderno** [comment]
With the PR I created, this is the result.

<img width="745" height="171" alt="Image" src="https://github.com/user-attachments/assets/25d8b674-0ea9-4c53-83a4-e7117a66e3e9" />

**Comment by swampopus** [comment]
I see you had to change the tests themselves to get the hashing concept to pass.

So, I did the same thing-- I just got rid of this test entirely:
```php
/**
   * Test JavaScript versioning.
   */
  function testVersionQueryString() {
    backdrop_add_js('core/misc/collapse.js', array('version' => 'foo'));
    backdrop_add_js('core/misc/ajax.js', array('version' => 'bar'));
    $javascript = backdrop_get_js();
    $this->assertTrue(strpos($javascript, 'core/misc/collapse.js?v=foo') > 0 && strpos($javascript, 'core/misc/ajax.js?v=bar') > 0 , t('JavaScript version identifiers correctly appended to URLs'));
  }
```
When I do that, my PR passes all tests, and works with only one line of code changed in common.inc.  See below result:

<img width="450" height="210" alt="Image" src="https://github.com/user-attachments/assets/7dbed7d4-fdf2-417f-bff4-f2a5250ed5be" />


Ultimately, It doesn't really matter to me which approach is taken, it just seems like all the extra hashing logic you're doing is just extra lines of code with very little benefit -- you have the core js files getting _unique_ cache-bust strings, instead of the `default_query_string` that all other contrib js files get.

But, now with two PRs that both achieve the same result, I'm happy to leave it up to the Backdrop admins to decide which approach they prefer (if either).

**Comment by avpaderno** [comment]
`::testVersionQueryString()` just verifies the expected version is attached to JavaScript URLs; it indeed needs to be changed when the string used as version is changed.  
Changing the test is different from removing it, though.

**Comment by swampopus** [comment]
> Changing the test is different from removing it, though.

In my PR, the versions are removed, so the test for the versions are also removed.

**Comment by quicksketch** [comment]
@avpaderno I like your PR approach but is it unnecessarily complicated? Would there be a serious downside to just using `$hash = hash('crc32', $item['version'] . $seed);` for all versions of PHP? Putting the seed in the hash like that seems like it would already be sufficient to prevent hash-table lookups.

**Comment by swampopus** [comment]
@avpaderno @quicksketch -- also curious-- why make the seed only 8 chars of the `backdrop_get_hash_salt()`?  Why not just use the entire salt, that's what it's there for, right?

**Comment by avpaderno** [comment]
The seed for xxHash and MurmurHash is expected to be an integer in the range from 0 to the platform defined `UINT_MAX`, usually 4294967295. That means a 8-digit hexadecimal number.
(That information is given in [Migrating from PHP 8.0.x to PHP 8.1.x / New Features](https://www.php.net/manual/en/migration81.new-features.php#migration81.new-features.hash).)

**Comment by avpaderno** [comment]
@quicksketch To simplify the code, it could be reduced to the following one.

```php
if (!empty($item['version'])) {
   $query_string = 'v=' . hash('crc32', $item['version'] . backdrop_get_hash_salt());
}
else {
  $query_string = $default_query_string;
}
```

xxHash has low collision rates, but that should not make any difference, when the input is a version string.

**Comment by swampopus** [comment]
I think that's the best way to go. I will close my PR.

**Comment by avpaderno** [comment]
I did a quick test by generating CRC32 values for version strings between 1.0.0 and 3.99.99 using `php -r '$salt = "55dbfec1cb6bb48aca247da587676dae0bcea812b527a4169bd4389e9d55e07b"; for ($a = 1; $a < 4; $a++) { for ($b = 0; $b < 100; $b++) { for ($c = 0; $c < 100; $c++) { $version = $a . "." . $b . "." . $c; echo hash("crc32", $version . $salt), " => ", $version, "\n"; } } }' > test_crc32.txt`. I then used `awk '{k=$1; v=$3; a[k]=(k in a ? a[k]" and "v : v); c[k]++} END{for(k in c) if(c[k]>1) print "CRC32 ["k"] generated by: "a[k]}' test_crc32.txt` to find duplicate CRC32 values. The last command just printed an empty line.

(I apologize: I was looking for a non-theoretical evidence that there can be no duplicates for values expected for a version string.)

**Comment by avpaderno** [comment]
This what the live preview shows now.

<img width="623" height="169" alt="Image" src="https://github.com/user-attachments/assets/e8453d74-2a5b-4406-9b2f-2fec55a11865" />

What could be further done is removing `v=` from the URL. That would remove any hint about us using version strings.  
There is no possibility of conflict between `?tjrvkm` and `?88f4b899`, for example, because `tjrvkm` is a Base36 number (six digits) while `88f4b899` is an hexadecimal number (eight digits). Even if `88f4b899` were accidentally truncated to `88f4b8` or `f4b899`, those values would be different.

**Comment by swampopus** [comment]
I think you're overthinking it ;)

You've struck oil-- you can stop drilling.

EDIT: But if you want to get rid of the v=, sounds fine.

**Comment by avpaderno** [comment]
I have updated the PR. `v=` is never added to URLs for JavaScript files. Now the change is just a line.

```php
$query_string = empty($item['version']) ? $default_query_string : hash('crc32', $item['version'] . backdrop_get_hash_salt());
```

**Comment by izmeez** [comment]
@swampopus You've been doing a lot in this issue, would you be willing to remove the "pr - needs code review" label and maybe add "Ready to be committed" ?

**Comment by swampopus** [comment]
Done and done!

**Comment by quicksketch** [comment]
I labeled this as a "bug" since we're putting it into a bug-fix release. Despite it being a security-hardening, not a true security bug. It's very low risk since the behavior/frequency of the query string changing is the same as before, and the query string structure now just matches between libraries that provide a `version` string and those that do not.

I merged https://github.com/backdrop/backdrop/pull/5315 into 1.x and 1.34.x. Thank you @avpaderno and @swampopus for taking the lead on this issue! Thanks @argiepiano, @olafgrabienski, and @izmeez for thoughtful review and feedback.
